File: app/src/main/java/com/highcom/passwordmemo/util/file/RestoreDbFile.java
package com.highcom.passwordmemo.util.file;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.provider.MediaStore;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.UiThread;
import androidx.annotation.WorkerThread;
import androidx.core.os.HandlerCompat;

import com.highcom.passwordmemo.R;
import com.highcom.passwordmemo.database.ListDataManager;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class RestoreDbFile {
    private Context context;
    private Activity activity;
    private RestoreDbFileListener listener;
    private Uri uri;
    private AlertDialog progressAlertDialog;
    private ProgressBar progressBar;

    public interface RestoreDbFileListener {
        void restoreComplete();
    }

    private class BackgroundTask implements Runnable {
        private final android.os.Handler _handler;

        public BackgroundTask(android.os.Handler handler) {
            _handler = handler;
        }

        @WorkerThread
        @Override
        public void run() {
            String destPath = context.getDatabasePath("PasswordMemoDB").getPath();
            File destFile = new File (destPath);
            destFile.delete();
            String srcPath = context.getDatabasePath("PasswordMemoDB_tmp").getPath();
            File srcFile = new File (srcPath);
            srcFile.renameTo(destFile);

            progressBar.setProgress(100);
            PostExecutor postExecutor = new PostExecutor();
            _handler.post(postExecutor);
        }
    }

    private class PostExecutor implements Runnable {
        @UiThread
        @Override
        public void run() {
            progressAlertDialog.dismiss();
            new AlertDialog.Builder(context)
                    .setTitle(context.getString(R.string.restore_db))
                    .setMessage(context.getString(R.string.db_restore_complete_message) + System.getProperty("line.separator") + getFileNameByUri(context, uri))
                    .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            listener.restoreComplete();
                        }
                    })
                    .show();
        }
    }

    public RestoreDbFile(Activity activity, RestoreDbFileListener listener) {
        this.activity = activity;
        this.context = activity;
        this.listener = listener;
    }

    public void restoreSelectFolder(final Uri uri) {
        this.uri = uri;
        String fileName = getFileNameByUri(context, uri);
        if (fileName.contains(".db")) {
            new AlertDialog.Builder(context)
                    .setTitle(context.getString(R.string.restore_db))
                    .setMessage(context.getString(R.string.restore_message_front) + getFileNameByUri(context, uri) + System.getProperty("line.separator") + context.getString(R.string.restore_message_rear))
                    .setPositiveButton(R.string.restore_button, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            if (restoreDatabase(uri)) {
                                execRestoreDatabase();
                            }
                        }
                    })
                    .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            String path = context.getDatabasePath("PasswordMemoDB_tmp").getPath();
                            File file = new File(path);
                            file.delete();
                        }
                    })
                    .show();
        } else {
            new AlertDialog.Builder(context)
                    .setTitle(context.getString(R.string.restore_db))
                    .setMessage(context.getString(R.string.restore_message_error_file))
                    .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            String path = context.getDatabasePath("PasswordMemoDB_tmp").getPath();
                            File file = new File(path);
                            file.delete();
                        }
                    })
                    .show();
        }
    }

    private boolean restoreDatabase(final Uri uri) {
        InputStream inputStream = null;
        try {
            inputStream = context.getContentResolver().openInputStream(uri);

            String path = context.getDatabasePath("PasswordMemoDB_tmp").getPath();
            File file = new File (path);
            OutputStream outputStream = new FileOutputStream(file);
            byte[] buf = new byte[1024];
            int len;
            while ((len = inputStream.read(buf)) > 0) {
                outputStream.write(buf, 0, len);
            }
            inputStream.close();
        } catch (Exception exc) {
            Toast ts = Toast.makeText(context, context.getString(R.string.db_restore_failed_message), Toast.LENGTH_SHORT);
            ts.show();
            return false;
        } finally {
            if (inputStream != null) {
                try {
                    inputStream.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        return true;
    }

    private void execRestoreDatabase()
    {
        new AlertDialog.Builder(context)
                .setTitle(context.getString(R.string.restore_db))
                .setMessage(context.getString(R.string.db_restore_confirm_message))
                .setPositiveButton(R.string.execute, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        // 取込み中のプログレスバーを表示する
                        progressAlertDialog = new AlertDialog.Builder(context)
                                .setTitle(R.string.db_restore_processing)
                                .setView(activity.getLayoutInflater().inflate(R.layout.alert_progressbar, null))
                                .create();
                        progressAlertDialog.show();
                        progressAlertDialog.setCancelable(false);
                        progressAlertDialog.setCanceledOnTouchOutside(false);
                        progressBar = progressAlertDialog.findViewById(R.id.ProgressBarHorizontal);

                        // ワーカースレッドで取込みを開始する
                        Looper mainLooper = Looper.getMainLooper();
                        Handler handler = HandlerCompat.createAsync(mainLooper);
                        BackgroundTask backgroundTask = new BackgroundTask(handler);
                        ExecutorService executorService  = Executors.newSingleThreadExecutor();
                        executorService.submit(backgroundTask);
                    }
                })
                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String path = context.getDatabasePath("PasswordMemoDB_tmp").getPath();
                        File file = new File (path);
                        file.delete();
                    }
                })
                .show();
    }

    private String getFileNameByUri(Context context, Uri uri) {
        String fileName = "";
        String[] projection = {MediaStore.MediaColumns.DISPLAY_NAME};
        Cursor cursor = context.getContentResolver()
                .query(uri, projection, null, null, null);
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                fileName = cursor.getString(
                        cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME));
            }
            cursor.close();
        }

        return fileName;
    }
}

File: app/src/main/java/com/highcom/passwordmemo/util/file/RestoreDbFile.kt
package com.highcom.passwordmemo.util.file

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import android.widget.ProgressBar
import android.widget.Toast
import androidx.annotation.UiThread
import androidx.annotation.WorkerThread
import androidx.core.os.HandlerCompat
import com.highcom.passwordmemo.R
import com.highcom.passwordmemo.data.PasswordMemoRoomDatabase
import com.highcom.passwordmemo.databinding.AlertProgressbarBinding
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.util.concurrent.Executors

/**
 * DBファイルの復元処理クラス
 *
 * @property activity ダイアログ表示用アクティビティ
 * @constructor
 * DBファイルの復元処理コンストラクタ
 *
 * @param listener 復元処理完了通知リスナー
 */
class RestoreDbFile(private val activity: Activity, listener: RestoreDbFileListener) {
    /** ダイアログ表示用コンテキスト */
    private val context: Context
    /** 復元処理完了通知リスナー */
    private val listener: RestoreDbFileListener
    /** DBファイル復元元ファイルURI */
    private var uri: Uri? = null
    /** プログレスダイアログ */
    private var progressAlertDialog: AlertDialog? = null
    /** プログレスバー */
    private var progressBar: ProgressBar? = null

    /**
     * DBファイル復元処理完了通知リスナークラス
     *
     */
    interface RestoreDbFileListener {
        /**
         * DBファイル復元処理完了通知処理
         *
         */
        fun restoreComplete()
    }

    /**
     * DBファイル復元用バックグラウンドタスク
     *
     * @property _handler 処理ハンドラ
     */
    private inner class BackgroundTask(private val _handler: Handler) : Runnable {
        /**
         * DBファイル復元実行処理
         *
         */
        @WorkerThread
        override fun run() {
            val destPath = context.getDatabasePath("PasswordMemoDB").path
            val destFile = File(destPath)
            destFile.delete()
            val srcPath = context.getDatabasePath("PasswordMemoDB_tmp").path
            val srcFile = File(srcPath)
            srcFile.renameTo(destFile)
            progressBar!!.progress = 100
            val postExecutor = PostExecutor()
            _handler.post(postExecutor)
        }
    }

    /**
     * バックグラウンド実行後のランナークラス
     *
     */
    private inner class PostExecutor : Runnable {
        /**
         * バックグラウンド実行後処理
         *
         */
        @UiThread
        override fun run() {
            progressAlertDialog!!.dismiss()
            AlertDialog.Builder(context)
                .setTitle(context.getString(R.string.restore_db))
                .setMessage(
                    context.getString(R.string.db_restore_complete_message) + System.getProperty(
                        "line.separator"
                    ) + getFileNameByUri(context, uri)
                )
                .setPositiveButton(R.string.ok) { _, _ -> }
                .setOnDismissListener { listener.restoreComplete() }
                .show()
        }
    }

    init {
        context = activity
        this.listener = listener
    }

    /**
     * DBファイル復元元フォルダ選択確認ダイアログ表示処理
     *
     * @param uri 復元元DBファイルURI
     */
    fun restoreSelectFolder(uri: Uri?) {
        this.uri = uri
        val fileName = getFileNameByUri(context, uri)
        if (fileName.contains(".db")) {
            AlertDialog.Builder(context)
                .setTitle(context.getString(R.string.restore_db))
                .setMessage(
                    context.getString(R.string.restore_message_front) + getFileNameByUri(
                        context,
                        uri
                    ) + System.getProperty("line.separator") + context.getString(R.string.restore_message_rear)
                )
                .setPositiveButton(R.string.restore_button) { _, _ ->
                    if (restoreDatabase(uri)) {
                        execRestoreDatabase()
                    }
                }
                .setNegativeButton(R.string.cancel) { _, _ ->
                    val path = context.getDatabasePath("PasswordMemoDB_tmp").path
                    val file = File(path)
                    file.delete()
                }
                .show()
        } else {
            AlertDialog.Builder(context)
                .setTitle(context.getString(R.string.restore_db))
                .setMessage(context.getString(R.string.restore_message_error_file))
                .setPositiveButton(R.string.ok) { _, _ ->
                    val path = context.getDatabasePath("PasswordMemoDB_tmp").path
                    val file = File(path)
                    file.delete()
                }
                .show()
        }
    }

    /**
     * DBファイル復元処理
     * * 指定されたパスのDBファイルを復元する
     *
     * @param uri 復元元DBファイルURI
     * @return 復元完了可否
     */
    private fun restoreDatabase(uri: Uri?): Boolean {
        var inputStream: InputStream? = null
        try {
            inputStream = context.contentResolver.openInputStream(uri!!)
            val path = context.getDatabasePath("PasswordMemoDB_tmp").path
            val file = File(path)
            val outputStream: OutputStream = FileOutputStream(file)
            val buf = ByteArray(1024)
            var len: Int
            while (inputStream!!.read(buf).also { len = it } > 0) {
                outputStream.write(buf, 0, len)
            }
            inputStream.close()
        } catch (exc: Exception) {
            val ts = Toast.makeText(
                context,
                context.getString(R.string.db_restore_failed_message),
                Toast.LENGTH_SHORT
            )
            ts.show()
            return false
        } finally {
            if (inputStream != null) {
                try {
                    inputStream.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        return true
    }

    /**
     * DBファイル復元実行処理
     * * DBファイル復元処理をバックグラウンドで実行する
     *
     */
    private fun execRestoreDatabase() {
        AlertDialog.Builder(context)
            .setTitle(context.getString(R.string.restore_db))
            .setMessage(context.getString(R.string.db_restore_confirm_message))
            .setPositiveButton(R.string.execute) { _, _ ->
                // リストアする前にDBを閉じる
                PasswordMemoRoomDatabase.closeDatabase()
                // 取込み中のプログレスバーを表示する
                val binding = AlertProgressbarBinding.inflate(activity.layoutInflater)
                progressAlertDialog = AlertDialog.Builder(context)
                    .setTitle(R.string.db_restore_processing)
                    .setView(binding.root)
                    .create()
                progressAlertDialog?.show()
                progressAlertDialog?.setCancelable(false)
                progressAlertDialog?.setCanceledOnTouchOutside(false)
                progressBar = binding.progressBarHorizontal

                // ワーカースレッドで取込みを開始する
                val mainLooper = Looper.getMainLooper()
                val handler = HandlerCompat.createAsync(mainLooper)
                val backgroundTask = BackgroundTask(handler)
                val executorService = Executors.newSingleThreadExecutor()
                executorService.submit(backgroundTask)
            }
            .setNegativeButton(R.string.cancel) { _, _ ->
                val path = context.getDatabasePath("PasswordMemoDB_tmp").path
                val file = File(path)
                file.delete()
            }
            .show()
    }

    /**
     * URIからファイルパス名を取得する処理
     *
     * @param context コンテキスト
     * @param uri ファイルURI
     * @return ファイルパス名
     */
    private fun getFileNameByUri(context: Context, uri: Uri?): String {
        var fileName = ""
        val projection = arrayOf(MediaStore.MediaColumns.DISPLAY_NAME)
        val cursor = context.contentResolver
            .query(uri!!, projection, null, null, null)
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                fileName = cursor.getString(
                    cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)
                )
            }
            cursor.close()
        }
        return fileName
    }
}